### PR TITLE
[TEP-0135] Update Affinity Assistant documentation

### PR DESCRIPTION
**Pull request by QuanZhang-William**
Part of [#6740][#6740]. This commit updates the Affinity Assistant related documentation to reflect recent changes in the Affinity Assistant modes. This commit also adds a chart summarizing how to configure the Affinity Assistant modes using `coschedule` and `disabled-affinity-assistant` feature flags.

/kind documentation

[#6740]: https://github.com/tektoncd/pipeline/issues/6740

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: docs/workspaces.md
Comment (lbernick) [inline docs/workspaces.md:383]:
```suggestion
**Note:** **coschedule pipelineruns** and **isolate pipelineruns** modes are still under development and not yet functional.
```

File: docs/workspaces.md
Comment (lbernick) [inline docs/workspaces.md:367]: Would it make sense to create a separate docs md file for the affinity assistant? it's a bit buried here

File: docs/workspaces.md
Comment (lbernick) [inline docs/workspaces.md:370]: worth noting that with this mode, you can only mount 1 pvc backed workspace to each taskrun

File: docs/workspaces.md
Comment (lbernick) [inline docs/workspaces.md:374]: can this be rewritten to focus less on implementation details? i.e. we really don't have to mention the affinity assistant at all; we can just say all pipelinerun pods get scheduled to the same node. (we can mention this uses inter-pod affinity even though it's an implementation detail as that may be relevant for some users due to performance concerns)

File: docs/workspaces.md
Comment (lbernick) [inline docs/workspaces.md:377]: should this be a description of the isolate-pipelineruns mode?

File: docs/workspaces.md
Comment (lbernick) [inline docs/workspaces.md:397]: it might also be helpful to include this writeup in addition to this table: https://github.com/tektoncd/community/pull/1025#discussion_r1231018163

File: config/config-feature-flags.yaml
Comment (lbernick) [inline config/config-feature-flags.yaml:43]: can you also link to affinity assistant docs here? https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#customizing-the-pipelines-controller-behavior

File: docs/affinityassistants.md
Comment (QuanZhang-William) [inline docs/affinityassistants.md:4]: I'm not sure why do we need it 😂

File: docs/workspaces.md
Comment (QuanZhang-William) [reply docs/workspaces.md]: Ahh sorry. Fix...

File: docs/workspaces.md
Comment (QuanZhang-William) [reply docs/workspaces.md]: SGTM, added in the new file

File: docs/workspaces.md
Comment (QuanZhang-William) [reply docs/workspaces.md]: Good catch, fixed.

File: config/config-feature-flags.yaml
Comment (QuanZhang-William) [reply config/config-feature-flags.yaml]: Sure, links added

File: docs/workspaces.md
Comment (QuanZhang-William) [reply docs/workspaces.md]: Added a paragraph for it, PTAL

File: docs/workspaces.md
Comment (QuanZhang-William) [reply docs/workspaces.md]: 👍 , I have rephrased the paragrah

File: docs/workspaces.md
Comment (QuanZhang-William) [reply docs/workspaces.md]: 👍 . I have added a new file for affinity assistant

File: docs/affinityassistants.md
Comment (lbernick) [inline docs/affinityassistants.md:34]:
```suggestion
In the future, the `disable-affinity-assistant` feature flag will be deprecated, and the Affinity Assistant Modes will be only determined by 
```

File: docs/affinityassistants.md
Comment (lbernick) [inline docs/affinityassistants.md:39]:
```suggestion
**Note:** For users who previously accepted the default behavior (enabling affinity assistant) but now want one of the new features, you need to set `disable-affinity-assistant` to "true" and then turn on the new behavior by setting the `coschedule` flag. For users who previously disabled the affinity assistant but want one of the new features, just set the `coschedule` flag accordingly.
```

File: docs/affinityassistants.md
Comment (lbernick) [inline docs/affinityassistants.md:41]: coscheduling = disabled is OK right?

File: docs/affinityassistants.md
Comment (lbernick) [inline docs/affinityassistants.md:106]:
```suggestion
significantly. We do not recommend using the affinity assistant in clusters larger than several hundred nodes.
```

File: docs/affinityassistants.md
Comment (lbernick) [inline docs/affinityassistants.md:8]: nit: can you make sure to use consistent capitalization for PVC, PipelineRun, and TaskRun?

**Review by lbernick — APPROVED** [review]

File: config/config-feature-flags.yaml
Comment (lbernick) [reply config/config-feature-flags.yaml]: sorry, I meant link to the new affinity assistant docs from https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#customizing-the-pipelines-controller-behavior, not the other way around.

Comment (QuanZhang-William) [reply config/config-feature-flags.yaml]: Oh sorry, just fixed

File: docs/affinityassistants.md
Comment (QuanZhang-William) [reply docs/affinityassistants.md:8]: Updated

File: docs/affinityassistants.md
Comment (QuanZhang-William) [reply docs/affinityassistants.md:41]: Yeah this is fine. And `disable-affinity-assistant` : true and `coschedule`: workspaces will also disabled AA during migration. I put it here just for simplicity as this feature will be completed soon and this note will be removed.

File: docs/affinityassistants.md
Comment (wlynch) [inline docs/affinityassistants.md:31]: "Incompatible" is a bit ambiguous here - could you explain this a bit more?

Typically when I think of incompatibility I expect either an error to be returned or one field to take precedence over the other. But in this case it doesn't sound like either is what happens, and Affinity Assistants do coexist with PodTemplates? (though I could be misunderstanding something)

File: docs/affinityassistants.md
Comment (wlynch) [inline docs/affinityassistants.md:39]: How can I figure out if I previously accepted the default behavior?

Comment (wlynch) [inline docs/affinityassistants.md:39]: Side note: it's a bit confusing that to turn on the coscheduling affinity assistant feature, you need to set disable-affinity-assistant=true - the flag names seem to conflict each other. idk if it's too late for this, but you may want to look into renaming these flags.

File: docs/affinityassistants.md
Comment (wlynch) [inline docs/affinityassistants.md:41]: You may want to make this warning more prominent - moving this up to the top + ⚠️ / 🚨 emojis + bigger font / bold would help.

We should also try and classify all of these with their stability level to set expectations around breaking changes (by default users will assume stable).

File: docs/affinityassistants.md
Comment (wlynch) [inline docs/affinityassistants.md:37]: This is the first mention of migration - when is this happening?

File: docs/affinityassistants.md
Comment (wlynch) [inline docs/affinityassistants.md:43]: As a user / someone who isn't really familiar with this feature, I don't know what this table is supposed to be telling me 😅 

Are these the config map values?

File: docs/affinityassistants.md
Comment (wlynch) [inline docs/affinityassistants.md:33]: Might be worth including the expected config map values somewhere and how they map to the modes above.

File: docs/affinityassistants.md
Comment (QuanZhang-William) [reply docs/affinityassistants.md:39]: Hi @wlynch. Yeah, there was quite a long discussion about the naming, default value and how to make the migration less confusing to users while maintaining backward compatibility: https://github.com/tektoncd/community/pull/1025#discussion_r1227749034 and https://github.com/tektoncd/community/pull/1031.

Based on the discussion, it appeared that turning `disable-affinity-assistant` on or off to honour `coschedule` is quite personal. This feature is added as an alpha feature, I think we can collect more user feedback before promoting it to beta.

File: docs/affinityassistants.md
Comment (QuanZhang-William) [reply docs/affinityassistants.md:21]: The implementation for this feature is about to be completed. Maybe we can merge this PR after the feature is fully supported and update this section.

/hold

File: docs/affinityassistants.md
Comment (QuanZhang-William) [reply docs/affinityassistants.md]: I have added some explanation for it, Affinity term in PodTemplate will be overwritten by Affinity Assistant if specified:

https://github.com/tektoncd/pipeline/blob/4f6c74a33b59d2bfc5d5ee14e2a9aa67d74b22f9/pkg/reconciler/taskrun/taskrun.go#L752C1-L756C3

File: docs/affinityassistants.md
Comment (QuanZhang-William) [reply docs/affinityassistants.md:35]: Added 9 month deprecation time to this part

File: docs/affinityassistants.md
Comment (QuanZhang-William) [reply docs/affinityassistants.md:37]: Yeah. I switched the order of the chart and the notes, so that the chart is now right below its explanations. PTAL

File: docs/affinityassistants.md
Comment (QuanZhang-William) [reply docs/affinityassistants.md:32]: Please see the above comment

File: docs/affinityassistants.md
Comment (QuanZhang-William) [reply docs/affinityassistants.md]: Good point! Will add the stability levels for different mode!

Comment (QuanZhang-William) [reply docs/affinityassistants.md]: Hi @wlynch. I have removed all the warnings and WIP in the documentation, and we will hold this PR until this feature is fully implemented (https://github.com/tektoncd/pipeline/pull/6927 is merged).

I have also added the stability levels for the feature, including adding it to the `alpha feature` list. PTAL 😄

**Review by vdemeester — APPROVED** [review]
/lgtm

**Comment by QuanZhang-William** [comment]
/hold cancel 

https://github.com/tektoncd/pipeline/pull/6927 is merged

**Pull request merged**
